src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java

```diff
@@ -744,12 +840,77 @@ private boolean handleMaterials(@NotNull final Block block, @NotNull final IBloc
             return true;
         }
 
-        if (isBlockFree(block, block.getMetaFromState(blockState)))
+        List<ItemStack> itemList = new ArrayList<>();
+        itemList.add(BlockUtils.getItemStackFromBlockState(blockState));
+        if(job instanceof JobBuilder && ((JobBuilder) job).getStructure() != null
+                && ((JobBuilder) job).getStructure().getBlockInfo() != null && ((JobBuilder) job).getStructure().getBlockInfo().tileentityData != null)
         {
-            return true;
+            itemList.addAll(getItemStacksOfTileEntity(((JobBuilder) job).getStructure().getBlockInfo().tileentityData));
+        }
+
+        for(ItemStack stack: itemList)
+        {
+            if(!isBlockFree(stack) && checkOrRequestItems(getTotalAmount(stack)))
+            {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    /**
+     * Check how much of a certain stuck is actually required.
+     * @param stack the stack to check.
+     * @return the new stack with the correct amount.
+     */
+    private ItemStack getTotalAmount(ItemStack stack)
+    {
+        AbstractBuildingWorker buildingWorker = getOwnBuilding();
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'buildingWorker' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)